module/owlib/src/c/ow_infernoembedded.c

```diff
@@ -0,0 +1,607 @@
+ */
+static GOOD_OR_BAD OW_device_info(const struct parsedname *pn, ie_device *device)
+{
+	BYTE write_string[] = { _1W_DEVICE};
+	BYTE buf[4];
+
+	memcpy(device->address, pn->sn, SERIAL_NUMBER_SIZE);
+
+	// Get the device type
+	struct transaction_log txn[] = {
+		TRXN_START,
+		TRXN_WRITE1(write_string),
+		TRXN_READ(buf, 4),
+		TRXN_END,
+	};
+
+	if (BAD(BUS_transaction(txn, pn))) {
+		return gbBAD;
+	}
+
+	device->device = read_uint32(buf);
+
+	// Get the device version
+	write_string[0] = _1W_VERSION;
+
+	if (BAD(BUS_transaction(txn, pn))) {
+		return gbBAD;
+	}
+
+	device->version = read_uint32(buf);
+
+	// Populate device specific info
+	switch (device->device) {
+	case RGBW_CONTROLLER:
+		return OW_rgbw_controller_info(pn, device);
+
+	case FIRMWARE_UPDATER:
+		return OW_firmware_updater_info(pn, device);
+
+	case SWITCH_MASTER:
+		return OW_switch_master_info(pn, device);
+
+	default:
+		LEVEL_DEBUG("Unknown device type %ld", device->device);
+		return gbBAD;
+	}
+}
+
+static GOOD_OR_BAD OW_ie_get_status(const struct parsedname *pn, uint64_t *status)
+{
+	BYTE write_string[] = { _1W_STATUS_REGISTER};
+	BYTE buf[8];
+
+	struct transaction_log t[] = {
+		TRXN_START,
+		TRXN_WRITE1(write_string),
+		TRXN_READ(buf, 8),
+		TRXN_END,
+	};
+
+	if (BAD(BUS_transaction(t, pn))) {
+		LEVEL_DEBUG("Status transaction failed");
+		return gbBAD;
+	}
+
+	*status = read_uint64(buf);
+
+	return gbGOOD;
+}
+
+
+/**
+ * Get metadata for a device
+ * @param pn the parsed device name
+ * @param device a pointer to a device that we will set (must be terminated with DEVICE_TERM)
+ * @return bad on error
+ */
+static GOOD_OR_BAD device_info(const struct parsedname *pn, ie_device **device)
+{
+	*device = NULL;
+
+	// Check the hash first
+#if OW_UTHASH
+	HASH_FIND(hh, device_cache, pn->sn, SERIAL_NUMBER_SIZE, *device);
+#endif
+
+	if (*device) {
+		LEVEL_DEBUG("Found device in hash");
+		// Found, all good
+		return gbGOOD;
+	}
+
+	// Not found, better fetch it from the bus
+	*device = new_device();
+	if (*device == NULL) {
+		LEVEL_DEBUG("Could not allocate new Inferno Embedded device");
+	}
+
+	if (BAD(OW_device_info(pn, *device))) {
+		DEVICE_TERM(*device);
+		return gbBAD;
+	}
+
+#if OW_UTHASH
+	LEVEL_DEBUG("Storing device in hash");
+	HASH_ADD( hh, device_cache, address, SERIAL_NUMBER_SIZE, *device);
+#endif
+
+	return gbGOOD;
+}
+
+/**
+ * Reboot the device into the firmware updater
+ * @param pn the parsed device name
+ */
+static GOOD_OR_BAD OW_ie_boot_firmware_updater(const struct parsedname *pn)
+{
+	BYTE write_string[] = { _1W_BOOT_FIRMWARE_UPDATER,  0};
+	BYTE crc = CRC8compute(pn->sn, 8, 0);
+	crc = CRC8compute(write_string, 1, crc);
+	write_string[1] = crc;
+
+	struct transaction_log t[] = {
+		TRXN_START,
+		TRXN_WRITE2(write_string),
+		TRXN_END,
+	};
+
+	if (BAD(BUS_transaction(t, pn))) {
+		return gbBAD;
+	}
+
+#if OW_UTHASH
+	ie_device *device;
+	if (BAD(device_info(pn, &device))) {
+		return gbBAD;
+	}
+	invalidate_device(&device);
+#endif
```

Comment:
Why is this block wrapped in 
```
#if OW_UTHASH
...
#endif 
```

Reply:
Hmm, the #ifdef in invalidate_device should wrap the whole function. There's little point in calling it if there is no cache to remove the device from.